virtcontainers/sandbox.go

```diff
@@ -1780,3 +1805,43 @@ func (s *Sandbox) AddDevice(info config.DeviceInfo) (api.Device, error) {
 
 	return b, nil
 }
+
+func (s *Sandbox) updateResources() error {
+	// the hypervisor.MemorySize is the amount of memory reserved for
+	// the VM and contaniners without memory limit
+
+	// Calculate Sandbox resources
+	sandboxCPUs := s.hypervisor.hypervisorConfig().NumVCPUs
+	sandboxMemoryByte := int64(s.hypervisor.hypervisorConfig().MemorySize) << utils.MibToBytesShift
+	for _, c := range s.containers {
+		sandboxMemoryByte += c.config.Resources.MemByte
+		sandboxCPUs += c.config.Resources.VCPUs
+	}
+
+	// Update VCPUs
+	s.Logger().WithField("cpus-sandbox", sandboxCPUs).Debugf("Request to hypervisor to update vCPUs")
+	oldCPUs, newCPUs, err := s.hypervisor.resizeVCPUs(sandboxCPUs)
+	if err != nil {
+		return err
+	}
+	// The CPUs were increased, ask agent to online them
+	if oldCPUs < newCPUs {
+		vcpusAdded := newCPUs - oldCPUs
+		if err := s.agent.onlineCPUMem(vcpusAdded, true); err != nil {
+			return err
+		}
+	}
+	s.Logger().Debugf("Sandbox CPUs: %d", newCPUs)
+
+	// Update Memory
+	s.Logger().WithField("memory-sandbox-size-byte", sandboxMemoryByte).Debugf("Request to hypervisor to update memory")
+	newMemory, err := s.hypervisor.resizeMemory(uint32(sandboxMemoryByte>>utils.MibToBytesShift), s.state.GuestMemoryBlockSizeMB)
```

Comment:
I think there is some problem. The reason we change the unit of c.config.Resources.Mem from MiB to Byte is to prevent the 4095B < memory < 1MiB from being lost and ignore 1B < memory < 4096B.

There is a example:
docker run 128M, we will hotplug 128M memory, and then,
docker update 128M+1B, we will do nothing,
docker update 100M+4096B, we will hotplug another 128M memory after calculate by `calcHotplugMemMiBSize`

By simply do 20 right shift to sandboxMemoryByte, we will lose memory less than 1M.

Reply:
Of course, in other architecture it may be not 4096B

Reply:
Sorry for the late response, I was on vacations, the only option I see is to move most of the logic bytes to avoid being converting in all the places.